opt: store and display mutation arbiter indexes in opt tests and EXPLAIN

mgartner opened the pull request:
#### opt: store mutation arbiter indexes and display in opt tests

This commit adds the list of arbiter indexes to opt test expression
trees. In order to facilitate this, the list of index ordinals that are
arbiters are stored on `MutationPrivate` for `Upsert` and `Insert`
expressions.

This commit also formats the canary column for `UPSERT`s like other
columns, instead of a plain column ID.

Additionally, the `optgen` error message produced when referencing a
type in an `*.opt` file that has not been registered is now less
ambiguous. With the previous error message it was unclear if it was
referencing `opt/metadata.go` or `optgen/metadata.go`.

Release justification: low-risk update to new functionality

Release note: None

#### opt: display arbiter indexes in EXPLAIN output

This commit adds a list of arbiter indexes to EXPLAIN output for UPSERT
and INSERT ON CONFLICT statements. These arbiters are used to detect
conflicts.

Release justification: low-risk update to new functionality

Release note (sql change): The output of EXPLAIN for UPSERT and INSERT
ON CONFLICT statements now includes  a list of arbiter indexes. These
arbiters are the indexes used for detecting conflicts between the insert
row and the existing rows in the table.

mgartner committed:
opt: store mutation arbiter indexes and display in opt tests

This commit adds the list of arbiter indexes to opt test expression
trees. In order to facilitate this, the list of index ordinals that are
arbiters are stored on `MutationPrivate` for `Upsert` and `Insert`
expressions.

This commit also formats the canary column for `UPSERT`s like other
columns, instead of a plain column ID.

Additionally, the `optgen` error message produced when referencing a
type in an `*.opt` file that has not been registered is now less
ambiguous. With the previous error message it was unclear if it was
referencing `opt/metadata.go` or `optgen/metadata.go`.

Release justification: low-risk update to new functionality

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/53172)
<!-- Reviewable:end -->

mgartner commented:
The first commit is from #53026 
The second commit is from #53171

RaduBerinde commented:
LGTM

I'd also plumb them through to the exec factory so they can be shown in the regular EXPLAIN. These changes can be backported after the branching though, they're very low risk.

mgartner committed:
opt: display arbiter indexes in EXPLAIN output

This commit adds a list of arbiter indexes to EXPLAIN output for UPSERT
and INSERT ON CONFLICT statements. These arbiters are used to detect
conflicts.

Release justification: low-risk update to new functionality

Release note (sql change): The output of EXPLAIN for UPSERT and INSERT
ON CONFLICT statements now includes  a list of arbiter indexes. These
arbiters are the indexes used for detecting conflicts between the insert
row and the existing rows in the table.

mgartner commented:
I've added the arbiters to EXPLAIN output. Is it odd that they are displayed in `EXPLAIN` and `EXPLAIN (OPT, VERBOSE)`, but not in `EXPLAIN (OPT)`? Happy to add them to `EXPLAIN (OPT)` if so.

RaduBerinde reviewed (approved):
Cool! I think it's fine to omit in non-verbose EXPLAIN (OPT).

mgartner commented:
bors r=RaduBerinde

mgartner commented:
bors r=RaduBerinde
